uber/okbuck
# Integration tests support

**Description**:
Added support for integration tests on java modules.

In order to enable it for your project, there are a few steps required:
1. Set `enableIntegrationTests` on `test` extension in okbuck
```groovy
okbuck {
    <... other configs ... >
    test {
        enableIntegrationTests = true
    }
}
```
2. Copy the file under `tooling/gradle/integration_tests.gradle` (of this project) to your own project
The file contents should have, at minimal, this format:
```groovy
sourceSets {
    integrationTest
}

// define custom test task for running integration tests
// This is used by okbuck in order to determine vm args and env variables
task integrationTest(type: Test) {
    description = 'Runs the integration tests.'
    group = 'verification'
    testClassesDirs = sourceSets.integrationTest.output.classesDirs
    classpath = sourceSets.integrationTest.runtimeClasspath
    outputs.upToDateWhen { false }
}
```
3. Apply file described at [1] on your module to create integration tests, right after applying your language plugin.
```groovy
apply plugin: "java-library"
apply from: rootProject.file("tooling/gradle/integration-tests.gradle")

<... rest of your configurations as they were before ...>
```
4. Now you can define the properties for `integrationTest` similar of how you'd define for `test`.
```
apply plugin: "java-library"
apply from: rootProject.file("tooling/gradle/integration-tests.gradle")

test {
    environment "TEST_ENV", "true"
    environment "CACHE_CASSANDRA_PORT", "1234"
    environment "CACHE_CASSANDRA_CONTACT_POINTS", "[testhost]"
    environment "OKBUCK_DATACENTER", "testdc1"
    environment "OKBUCK_PORT_HTTP", "4321"
}

integrationTest {
    environment "ANOTHER", "true"
}

dependencies {
    api deps.external.gson
    implementation deps.external.dagger

    annotationProcessor deps.apt.javax
    annotationProcessor deps.apt.daggerCompiler

    compileOnly deps.androidx.annotations

    testImplementation deps.test.junit

    integrationTestImplementation deps.test.junit
    integrationTestAnnotationProcessor deps.apt.daggerCompiler
    integrationTestImplementation deps.external.avroIpc
}
```

**Caveats**
Ideally we would want to not have steps two and three. That would require us to be able to map the configuration under that file into calling the correct APIs inside the plugin and adding them programatically there, jumping directly for step 4. I could not find an easy/straight forward way to do that, so I'm open for suggestions.
